### ISSUE TITLE
Cannot repeat experiment, even with same rng_seed

### ISSUE THREAD

**Issue by ilija139** [NervanaSystems/neon#193]
I want to be able to reproduce the experiments that I run. 
For example if I run multiple times: 

`./examples/cifar10_conv.py -r 139` 

I expect the Misclassification error and the cost per epoch to be always the same, since I'm using the same `rng_seed`. However this is not the case. 
Am I doing something wrong or? 

Edit: I'm able to reproduce results with `mnist_mlp.py` but not with `cifar10_conf.py`

**Comment by nervetumer** [comment]
Some speed optimizations can lead to non-deterministic behavior.  When running on GPU, this model uses a convolution and batchnorm implementations that require atomic addition operations.  The non-determinisim here arises from the numerical differences causes by running these operations in a different order.  We are going to push a change soon that will make the convolution operations not use atomics by default.  We are looking into a similar change for the batchnorm operations.

**Comment by ilija139** [comment]
Is there a way to turn off the use of atomics with the current version?

**Comment by nervetumer** [comment]
For the convolution layer you can set the "deterministic_update" parameter in the nervanagpu.py file to True.  For batch norm it may be a little more complicated and the solution is not tested yet.  You can try to set the sum parameter sum to be False.  For macro layers like Conv and Affine, you'll need to keep batch_norm active but not have it pass bsum=True to the Convolution or Linear layer.

**Comment by ilija139** [comment]
Ok, ICML deadline is over, so I had time to test this. And it **works**, thanks a lot!

Modifications needed:

1. In `neon/backends/nervanagpu.py`, line 665, change the default value of `deterministic_update` to `False`.
2. In `neon/layers/layer.py` line 1130 (`Conv` constructor) set `bsum=False` when appending `Convolution` layer. 
3. In `neon/layers/layer.py` line 1092 (`Affine` constructor) set `bsum=False` when appending `Linear` layer. 

After this changes CNN with batch norm are deterministic. 
Thanks

**Comment by ursk** [comment]
Shouldn't that be `deterministic_update` to `True`?
